inspector: added --inspect-publish-uid

ak239 opened the pull request:
This flag specifies how inspector websocket url should be exposed.
Supported options:
- stderr - reports websocket as a message to stderr,
- http - exposes /json/list endpoint that contains inspector websocket
  url,
- binding - `require('inspector').url()`.

Related discussion: https://github.com/nodejs/diagnostics/issues/303

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

eugeneo commented on src/inspector_socket_server.cc:
```diff
@@ -94,14 +94,20 @@ const char* MatchPathSegment(const char* path, const char* expected) {
-  const char HEADERS[] = "HTTP/1.0 200 OK\r\n"
+void SendHttpResponse(InspectorSocket* socket,
+                      const std::string& response,
+                      int code = 200) {
```
Please, don't use parameter default values.

ak239 replied:
done.

eugeneo commented on src/inspector_io.h:
```diff
@@ -48,7 +48,8 @@ class InspectorIo {
       const std::string& path,
-      std::shared_ptr<HostPort> host_port);
+      std::shared_ptr<HostPort> host_port,
+      const std::string& inspect_publish_uid);
```
I believe that this class should not be parsing CLI parameter values but take two booleans instead. I would suggest parsing them as close to the node_options.cc as possible.

ak239 replied:
I moved parsing to node_options.cc, since we need to parse it there in any case to validate argument value.

eugeneo commented on src/inspector_socket_server.cc:
```diff
@@ -241,11 +251,26 @@ void PrintDebuggerReadyMessage(
   state_ = ServerState::kNew;
+  std::vector<std::string> uid_destinations =
+      SplitString(inspect_publish_uid, ',');
+  for (const std::string& destination : uid_destinations) {
```
Please move this to `node_options.cc` if possible. If not possible - do this in `inspector_agent.cc`.

ak239 replied:
done.

eugeneo commented on src/inspector_socket_server.cc:
```diff
@@ -278,7 +303,8 @@ void InspectorSocketServer::SessionTerminated(int session_id) {
   connected_sessions_.erase(session_id);
   if (connected_sessions_.empty()) {
-    if (was_attached && state_ == ServerState::kRunning
+    if (publish_uid_stderr_
```
I suggest passing this all the way into `PrintDebuggerReadyMessage`

ak239 replied:
done.

ak239 commented:
I addressed comments and added another option - `binding`. When this option is not presented, `require('inspector').url()` returns empty url.

addaleax commented on src/node_options.cc:
```diff
@@ -44,6 +44,24 @@ void DebugOptions::CheckOptions(std::vector<std::string>* errors) {
   }
+
+  if (inspect_publish_uid_string.empty())
+    inspect_publish_uid_string = "stderr,http,binding";
```
Can you make this the default value in `node_options.h`? That way an empty string can be passed intentionally

ak239 replied:
done.

addaleax commented on src/inspector_js_api.cc:
```diff
@@ -261,7 +261,7 @@ void Url(const FunctionCallbackInfo<Value>& args) {
   InspectorIo* io = agent->io();
 
-  if (!io) return;
+  if (!io || !io->publish_uid_binding()) return;
```
Is this technically a breaking change?

ak239 replied:
as long as I passed "binding" by default - it is not a breaking change but I removed binding from this PR.

eugeneo commented:
> I addressed comments and added another option - binding. When this option is not presented, require('inspector').url() returns undefined.

IMHO, that's unnecessary. Code that can access inspector API can already do pretty much anything.

ak239 commented:
> > I addressed comments and added another option - binding. When this option is not presented, require('inspector').url() returns undefined.
> 
> IMHO, that's unnecessary. Code that can access inspector API can already do pretty much anything.

I removed `binding` from this PR and I'd like to create separate one and when `binding` is not passed - disable inspector JS bindings all together. WDYT?

eugeneo commented:
> I removed binding from this PR and I'd like to create separate one and when binding is not passed - disable inspector JS bindings all together. WDYT?

I think that would warrant more discussions. I do not see utility of disabling the bindings. It would not help with the security (JS bindings are only accessible from inside the Node runtime).

My main concern is that people would disable JS bindings "for security" and end up confused why some ecosystem tools stop working.

ak239 commented:
> I think that would warrant more discussions. I do not see utility of disabling the bindings. It would not help with the security (JS bindings are only accessible from inside the Node runtime).

Imagine for a second compromised `superpopular` package that uses `inspector` to get anything from any package around violating any module/function scopes. It can access any other package information, it can open inspector websocket url and send it somewhere.

In general if I can inject some code to Node process using compromised dependency or some other way, I get access to part of JS heap but not to everything. As soon as I can use inspector in this snippet - I get access to everything.

> My main concern is that people would disable JS bindings "for security" and end up confused why some ecosystem tools stop working.

If people need some tool to work then they need to allow inspection. And some ecosystem tool can easily check that inspection flag was not passed and ask user to pass it - it is not a big deal.

Trott commented:
@eugeneo @ak239 Can/should this land? Or do you two still need to work out stuff?

ak239 commented:
I need to fix a test first.

Trott commented:
> I need to fix a test first.

OK, thanks! I've added the `blocked` label. (When the test is fixed, remove it!)

ak239 commented:
Landed in f0018a5

BridgeAR commented on doc/api/cli.md:
```diff
@@ -980,6 +987,7 @@ Node.js options that are allowed are:
 - `--inspect`
 - `--inspect-brk`
 - `--inspect-port`
+- `--inspect-publish-uid`
```
This should also be documented in `./doc/node.1`.

BridgeAR commented:
Should this actually be semver-minor?

ak239 commented:
@BridgeAR this pull request does not change default behavior so I believe we do not need `semver-minor` for it.

BridgeAR commented:
@ak239 if it would change the default, I'd expect it to be semver-major? We normally declare all new features as semver-minor.

ak239 commented:
@BridgeAR I see, thank you for clarification. It is new feature so `semver-minor` sounds good to me.
